File: jscomp/others/bs_LinkList.ml


type 'a cell = {
  mutable head : 'a;
  mutable tail : 'a opt_cell
}

and 'a opt_cell = 'a cell Js.null

and 'a t = {
  length : int ;
  data : 'a opt_cell
} [@@bs.deriving abstract]


external assertAsNonNull : 'a Js.null -> 'a = "%identity"
external tailOption : 'a cell -> 'a cell option = "tail" [@@bs.get] [@@bs.return null_to_opt] 

let toOpt  =  Js.nullToOption
let return = Js.Null.return 
let empty = Js.Null.empty

let headOpt ( x  : _ t) = 
  toOpt (data x)

let tailOpt (x : _ t) =
  match toOpt (data x ) with 
  | None -> None
  | Some x -> tailOption x 

let rec lengthCellAux (x : _ opt_cell) acc =   
  match toOpt x with 
  | None -> acc 
  | Some x -> lengthCellAux (tail x) (acc + 1)

let checkInvariant (x : _ t) : unit =  
  [%assert length x = lengthCellAux ( data x ) 0]

let rec nextAuxAssert (opt_cell : 'a opt_cell) n =   
  let cell = (assertAsNonNull opt_cell) in 
  if n = 0 then 
    (head cell)
  else 
    nextAuxAssert (tail cell) (n - 1)

let nthOpt x n =   
  if n < 0 then None 
  else if n < (length x) then  
    Some (nextAuxAssert  (data x) n)
  else 
    None

let nthAssert  x n =
  if n < 0 then [%assert "Neg"]
  else nextAuxAssert (data x) n 

let rec copyAux (cellX : _ opt_cell)  (prec : _ cell) = 
  match toOpt cellX with 
  | None -> prec
  | Some cellX -> 
    let h, t = head cellX, tail cellX in 
    let next = cell ~head:h ~tail:empty in 
    tailSet prec (return next);
    copyAux t next 

let copyNonEmptyTo xs  ys = 
  let cell = cell ~head:(head xs) ~tail:empty in 
  let newTail = copyAux (tail xs) cell in 
  tailSet newTail ys; 
  cell 

let append (x : 'a t) (y : 'a t) : 'a t =      
  let lenX = length x in  
  if lenX = 0 then y 
  else 
    let lenY = length y in 
    if lenY = 0 then x
    else 
      let h =  assertAsNonNull (data x)  in 
      (* let cell = cell ~head:(head h) ~tail:empty in 
      let newTail = copyAux (tail h) cell  in 
      tailSet newTail (data y) ;  *)
      let cell = copyNonEmptyTo h (data y) in 
      t ~length:(lenX + lenY) ~data:(return cell )


let init n f = 
  if n < 0 then [%assert "Invalid_argument"]
  else 
  if n = 0 then 
    t ~length:0 ~data:empty (* TODO could be shared *)
  else 
    let headX = (cell ~head:(f 0 [@bs]) ~tail:empty) in 
    let  cur = ref headX in 
    let i = ref 1 in 
    while !i < n do 
      let v = cell ~head:(f !i [@bs]) ~tail:empty in 
      tailSet !cur (return v);
      cur := v;
      incr i; 
    done ;
    t ~length:n ~data:(return headX)

let rec fillAux arr i (cell_opt : _ opt_cell) =       
  match toOpt cell_opt with 
  | None -> ()
  | Some x ->
    Bs_Array.unsafe_set arr i (head x) ; 
    fillAux arr (i + 1) (tail x)

let toArray (x : _ t) = 
  let len = length x in 
  let arr = Bs.Array.makeUninitializedUnsafe len in 
  fillAux arr 0 (data x);
  arr

File: lib/js/bs_LinkList.js
'use strict';

var Js_primitive = require("./js_primitive.js");

function toOpt(prim) {
  if (prim === null) {
    return /* None */0;
  } else {
    return [prim];
  }
}

function $$return(prim) {
  return prim;
}

var empty = null;

function headOpt(x) {
  return Js_primitive.null_to_opt(x.data);
}

function tailOpt(x) {
  var match = x.data;
  if (match !== null) {
    return Js_primitive.null_to_opt(match.tail);
  } else {
    return /* None */0;
  }
}

function lengthCellAux(_x, _acc) {
  while(true) {
    var acc = _acc;
    var x = _x;
    if (x !== null) {
      _acc = acc + 1 | 0;
      _x = x.tail;
      continue ;
      
    } else {
      return acc;
    }
  };
}

function checkInvariant(x) {
  if (x.length !== lengthCellAux(x.data, 0)) {
    throw new Error("File \"bs_LinkList.ml\", line 37, characters 4-10");
  } else {
    return 0;
  }
}

function nextAuxAssert(_opt_cell, _n) {
  while(true) {
    var n = _n;
    var opt_cell = _opt_cell;
    if (n) {
      _n = n - 1 | 0;
      _opt_cell = opt_cell.tail;
      continue ;
      
    } else {
      return opt_cell.head;
    }
  };
}

function nthOpt(x, n) {
  if (n < 0 || n >= x.length) {
    return /* None */0;
  } else {
    return /* Some */[nextAuxAssert(x.data, n)];
  }
}

function nthAssert(x, n) {
  if (n < 0) {
    throw new Error("Neg");
  } else {
    return nextAuxAssert(x.data, n);
  }
}

function copyAux(_cellX, _prec) {
  while(true) {
    var prec = _prec;
    var cellX = _cellX;
    if (cellX !== null) {
      var h = cellX.head;
      var t = cellX.tail;
      var next = {
        head: h,
        tail: empty
      };
      prec.tail = next;
      _prec = next;
      _cellX = t;
      continue ;
      
    } else {
      return prec;
    }
  };
}

function copyNonEmptyTo(xs, ys) {
  var cell = {
    head: xs.head,
    tail: empty
  };
  var newTail = copyAux(xs.tail, cell);
  newTail.tail = ys;
  return cell;
}

function append(x, y) {
  var lenX = x.length;
  if (lenX) {
    var lenY = y.length;
    if (lenY) {
      var h = x.data;
      var cell = copyNonEmptyTo(h, y.data);
      return {
              length: lenX + lenY | 0,
              data: cell
            };
    } else {
      return x;
    }
  } else {
    return y;
  }
}

function init(n, f) {
  if (n < 0) {
    throw new Error("Invalid_argument");
  } else if (n) {
    var headX = {
      head: f(0),
      tail: empty
    };
    var cur = headX;
    var i = 1;
    while(i < n) {
      var v = {
        head: f(i),
        tail: empty
      };
      cur.tail = v;
      cur = v;
      i = i + 1 | 0;
    };
    return {
            length: n,
            data: headX
          };
  } else {
    return {
            length: 0,
            data: empty
          };
  }
}

function fillAux(arr, _i, _cell_opt) {
  while(true) {
    var cell_opt = _cell_opt;
    var i = _i;
    if (cell_opt !== null) {
      arr[i] = cell_opt.head;
      _cell_opt = cell_opt.tail;
      _i = i + 1 | 0;
      continue ;
      
    } else {
      return /* () */0;
    }
  };
}

function toArray(x) {
  var len = x.length;
  var arr = new Array(len);
  fillAux(arr, 0, x.data);
  return arr;
}

exports.toOpt          = toOpt;
exports.$$return       = $$return;
exports.empty          = empty;
exports.headOpt        = headOpt;
exports.tailOpt        = tailOpt;
exports.lengthCellAux  = lengthCellAux;
exports.checkInvariant = checkInvariant;
exports.nextAuxAssert  = nextAuxAssert;
exports.nthOpt         = nthOpt;
exports.nthAssert      = nthAssert;
exports.copyAux        = copyAux;
exports.copyNonEmptyTo = copyNonEmptyTo;
exports.append         = append;
exports.init           = init;
exports.fillAux        = fillAux;
exports.toArray        = toArray;
/* empty Not a pure module */
